src/tokens/tokens.service.ts

```diff
@@ -0,0 +1,398 @@
+          to: dto.to,
+        },
+        this.postOptions(dto.operator, dto.requestId),
+      ),
+    );
+    return { id: response.data.id };
+  }
+
+  async transfer(dto: TokenTransfer): Promise<AsyncResponse> {
+    const response = await lastValueFrom(
+      this.http.post<EthConnectAsyncResponse>(
+        `${this.contractInstanceUrl}/${dto.poolId}/transferWithData`,
+        {
+          from: dto.from,
+          to: dto.to,
+          amount: dto.amount,
+          data: encodeHex(dto.data ?? ''),
+        },
+        this.postOptions(dto.operator, dto.requestId),
+      ),
+    );
+    return { id: response.data.id };
+  }
+
+  async burn(dto: TokenBurn): Promise<AsyncResponse> {
+    const response = await lastValueFrom(
+      this.http.post<EthConnectAsyncResponse>(
+        `${this.contractInstanceUrl}/${dto.poolId}/burnWithData`,
+        {
+          data: encodeHex(dto.data ?? ''),
+          from: dto.from,
+          amount: dto.amount,
+        },
+        this.postOptions(dto.operator, dto.requestId),
+      ),
+    );
+    return { id: response.data.id };
+  }
+
+  async balance(dto: TokenBalanceQuery): Promise<TokenBalance> {
+    const response = await lastValueFrom(
+      this.http.get<EthConnectReturn>(`${this.contractInstanceUrl}/${dto.poolId}/balanceOf`, {
+        params: {
+          account: dto.account,
+        },
+        ...basicAuth(this.username, this.password),
+      }),
+    );
+
+    return { balance: response.data.output };
+  }
+
+  async getReceipt(id: string): Promise<EventStreamReply> {
+    const response = await lastValueFrom(
+      this.http.get<EventStreamReply>(`${this.baseUrl}/reply/${id}`, {
+        validateStatus: status => status < 300 || status === 404,
+        ...basicAuth(this.username, this.password),
+      }),
+    );
+    if (response.status === 404) {
+      throw new NotFoundException();
+    }
+    return response.data;
+  }
+}
+
+class TokenListener implements EventListener {
+  private readonly logger = new Logger(TokenListener.name);
+
+  private uriPattern: string | undefined;
+
+  constructor(
+    private http: HttpService,
+    private instanceUrl: string,
+    private topic: string,
+    private username: string,
+    private password: string,
+  ) {}
+
+  async onEvent(subName: string, event: Event, process: EventProcessor) {
+    switch (event.signature) {
+      case tokenCreateEventSignature:
+        process(this.transformTokenCreateEvent(subName, event));
+        break;
+      case transferEventSignature:
+        process(await this.transformTransferEvent(subName, event));
+        break;
+      default:
+        this.logger.error(`Unknown event signature: ${event.signature}`);
+        return undefined;
+    }
+  }
+
+  private transformTokenCreateEvent(
+    subName: string,
+    event: TokenCreateEvent,
+  ): WebSocketMessage | undefined {
+    const { data } = event;
+    const decodedData = decodeHex(data.data ?? '');
+
```

Comment:
I think you should unpack the subName here and:
1. if it's the "base" subscription, deliver all token pool creation events
2. if it's a pool-specific subscription, only deliver the creation event for _that_ pool (and drop all others)